### ISSUE TITLE
`hintedhandoff_additional_test.TestHintedHandoff.test_hintedhandoff_sync_point_api` failure: sync_point DONE instead of IN_PROGRESS; and hang during shutdown

### ISSUE THREAD

**Issue by kbr-scylla** [scylladb/scylladb#18733]
Failure spotted in dtest-debug job
https://jenkins.scylladb.com/job/scylla-master/job/dtest-debug/295/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split006___test_hintedhandoff_sync_point_api/

https://jenkins.scylladb.com/job/scylla-master/job/dtest-debug/295/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split006___test_hintedhandoff_sync_point_api_2/

The two jenkins pages refer to the same test, but different causes of failure.

One failure was because there was an assertion failure
```python3
        #####
        logger.debug("SUBTEST 5: Create a hint sync point and decommission the target node - waiting should succeed")
    
        logger.debug("Stopping node2...")
        node2.stop(wait_other_notice=True)
    
        logger.debug("Pause hint replay on node1")
        self.enable_error("hinted_handoff_pause_hint_replay", node1)
    
        logger.debug(f"Inserting {len(keys5)} keys...")
        insert_c1c2(session, keys=keys5, consistency=ConsistencyLevel.ANY)
    
        logger.debug("Starting node2...")
        node2.start(wait_other_notice=True)
    
        logger.debug("Create hint sync point on node1")
        sync_point_id = create_sync_point(node=node1)
    
        logger.debug("Decommissioning node2...")
        node2.decommission()
    
        logger.debug("Check that the sync point is not yet resolved (because hint replay is paused)")
>       wait_for_sync_point(sync_point_id, 0, expect="IN_PROGRESS")
```

```
>           assert status.json() == expect
E           AssertionError: assert 'DONE' == 'IN_PROGRESS'
E             - IN_PROGRESS
E             + DONE
```

The second failure was because the dtest framework found an unexpected coredump from node1 on shutdown.
The logs of node1 at the end look like this:
```
INFO  2024-05-12 06:56:34,529 [shard 1:main] task_manager - Unregistered module repair
INFO  2024-05-12 06:56:34,530 [shard 0:main] init - Shutting down repair service was successful
INFO  2024-05-12 06:56:34,530 [shard 0:main] init - Shutting down drain storage proxy
INFO  2024-05-12 06:56:34,530 [shard 0:main] hints_manager - Asked to stop a shard hint manager
INFO  2024-05-12 06:56:34,530 [shard 0:main] hints_manager - Asked to stop a shard hint manager
INFO  2024-05-12 06:56:34,530 [shard 0:main] hints_manager - Shard hint manager has stopped
INFO  2024-05-12 06:56:34,531 [shard 1:main] hints_manager - Asked to stop a shard hint manager
INFO  2024-05-12 06:56:34,531 [shard 1:main] hints_manager - Asked to stop a shard hint manager
INFO  2024-05-12 06:56:34,531 [shard 1:main] hints_manager - Shard hint manager has stopped
```

And dtest-gw0.log says this:
```
07:06:32,397 750     ccm                            WARNING  cluster.py          :760  | test_hintedhandoff_sync_point_api: node1: node1 is still running after 600 seconds. Trying to generate coredump using kill(58388, SIGQUIT)...
07:06:33,398 750     ccm                            WARNING  cluster.py          :760  | test_hintedhandoff_sync_point_api: node1: node1 scylla-jmx is still running. Killing process using kill(600, SIGKILL)...
07:06:33,400 750     dtest_setup                    DEBUG    dtest_setup.py      :628  | test_hintedhandoff_sync_point_api: exclude_errors: []
07:06:33,407 750     dtest_setup                    DEBUG    dtest_setup.py      :628  | test_hintedhandoff_sync_point_api: exclude_errors: []
07:06:33,432 750     dtest_setup                    WARNING  dtest_setup.py      :164  | test_hintedhandoff_sync_point_api: Moving core file /jenkins/workspace/scylla-master/dtest-debug/scylla-dtest/scylla.58388.1715497592.core to /jenkins/workspace/scylla-master/dtest-debug/scylla-dtest/logs-full.debug.006/1715497593421_hintedhandoff_additional_test.py::TestHintedHandoff::test_hintedhandoff_sync_point_api/node1-scylla.58388.1715497592.core
```

So what happened is that shutdown of the node hanged (somewhere inside `hints_manager`), dtest framework timed out, told the process to generate a coredump, and then killed it --- and then failed because coredump was found. (The failure report is not perfect, dtest framework should say that it timed out waiting for node shutdown, but I digress)

GitHub doesn't allow me to upload the coredump because it's over 25MB (even after compression). Could be useful to debug the issue though, to understand where the process hanged -- so whoever sees this issue I recommend downloading the coredump locally before artifacts get cleared.

Anyway uploading logs:
[dtest-gw0.log](https://github.com/scylladb/scylladb/files/15351636/dtest-gw0.log)
[node1.log](https://github.com/scylladb/scylladb/files/15351638/node1.log)
[node2.log](https://github.com/scylladb/scylladb/files/15351639/node2.log)

Could it be related to the recent huge change in hints? 64ba620dc299365deaf4e9d369b2982f09c614c2

**Comment by dawmd** [comment]
I don't know yet why the test failed, but I have a hypothesis as to why stopping the node hung.

---

The scenario of the subtest that failed is this:

1. Have a cluster of two nodes – node 1 and node 2.
2. Stop node 2.
3. Pause hint replay on node 1 by error injection.
4. Save hints on node 1 towards node 2.
5. Start node 2 again.
6. Create a sync point on node 1.
7. Decommission node 2.
8. Await the sync point on node 1 (**this step fails!**).

---

There are two crucial points here:

(A) Node 1 has still the error injection enabled when the test fails preventing it from sending hints, cf. *Listing 1* below.
(B) Node 1 is draining hints towards node 2 because it was decommissioned.

They are the key to understand the problem.

---

The timeline, according to my hypothesis, looked like this:

1. When we decommissioned node 2, node 1 started draining hints towards it. Among the things that happened is we took `db::hints::manager`'s gate to only release it when the draining finished.
2. Because node 1 had the error injection enabled, it wasn't able to send any hint, so it kept trying ending up sleeping every time it attempted to do so, cf. *Listing 1*.
3. The test now fails (for whatever reason) leading it to stop node 1. `db::hints::manager::stop()` is called. One of the first things it does is wait for its gate to close – the same gate it took for draining. Draining is still happening, so the gate is not going to be released, leading to an infinite loop where the hint sender keeps sleeping hoping to send the hints and the hint manager waiting for it to finish.

The reason why the gate will never be released lies, again, in the error injection enabled. The first `if` statement in *Listing 1* will always resolve to `true` because we're draining the hints. The second `if` will *most likely* always resolve to `false`. The "most likely" in the previous statement corresponds to my current (and fuzzy) understanding of topology changes. I guess that node 2 is no longer present in the `locator::topology_metadata` at that point. If we assume that, `db::hints::internal::hint_sender::can_send()` should always return `true` (cf. *Listing 2*) because `maybe_ep` will be an empty optional and `tmptr->is_normal_token_owner(_ep_key)` will resolve to `false`.

We're now in an infinite loop from `Listing 1* where the two `if`s that should break it have become useless because we're in the middle of the draining process and the error injection keeps it looping.

---

If my hypothesis is correct, it doesn't look like a bug that could affect any user of the database. It could only happen because of the two conditions – that I named (A) and (B) – taking place at the same time.

I don't have time to prove or disprove this theory today, but I'll try to reproduce the issue tomorrow and report my findings.

---

<details>
<summary>Listings</summary>

*Listing 1*:
```cpp
/// Fragment of `bool hint_sender::send_one_file(const sstring& fname)`.

while (true) {
    // Check that we can still send the next hint. Don't try to send it if the destination host
    // is DOWN or if we have already failed to send some of the previous hints.
    if (!draining() && ctx_ptr->segment_replay_failed) {
        co_return;
    }

    // Break early if stop() was called or the destination node went down.
    if (!can_send()) {
        ctx_ptr->segment_replay_failed = true;
        co_return;
    }

    co_await flush_maybe();

    if (utils::get_local_injector().enter("hinted_handoff_pause_hint_replay")) {
        // We cannot send the hint because hint replay is paused.
        // Sleep 100ms and do the whole loop again.
        //
        // Jumping to the beginning of the loop makes sure that
        // - We regularly check if we should stop - so that we won't
        //   get stuck in shutdown.
        // - flush_maybe() is called regularly - so that new segments
        //   are created and we help enforce the "at most 10s worth of
        //   hints in a segment".
        co_await sleep(std::chrono::milliseconds(100));
        continue;
    } else {
        co_await send_one_hint(ctx_ptr, std::move(buf), rp, secs_since_file_mod, fname);
        break;
    }
};
```

*Listing 2*:
```cpp
bool hint_sender::can_send() noexcept {
    if (stopping() && !draining()) {
        return false;
    }

    const auto tmptr = _shard_manager._proxy.get_token_metadata_ptr();
    const auto maybe_ep = std::invoke([&] () noexcept -> std::optional<gms::inet_address> {
        try {
            return tmptr->get_endpoint_for_host_id_if_known(_ep_key);
        } catch (...) {
            return std::nullopt;
        }
    });

    try {
        // `hint_sender` can never target this node, so if the returned optional is empty,
        // that must mean the current locator::token_metadata doesn't store the information
        // about the target node.
        if (maybe_ep && _gossiper.is_alive(*maybe_ep)) {
            _state.remove(state::ep_state_left_the_ring);
            return true;
        } else {
            if (!_state.contains(state::ep_state_left_the_ring)) {
                _state.set_if<state::ep_state_left_the_ring>(!tmptr->is_normal_token_owner(_ep_key));
            }
            // If the node is not part of the ring, we will send hints to all new replicas.
            // Note that if the optional -- `maybe_ep` -- is empty, that could mean that `_ep_key`
            // is the locator::host_id of THIS node. However, that's impossible because instances
            // of `hint_sender` are only created for OTHER nodes, so this logic is correct.
            return _state.contains(state::ep_state_left_the_ring);
        }
    } catch (...) {
        return false;
    }
}
```
</details>

CC @piodul @kbr-scylla

**Comment by dawmd** [comment]
The bug appeared in the CI of scylladb/scylla-dtest#4717. Because the patch was adding additional logs, we've finally obtained some more information about the internal doings of Scylla.

Jenkins job: https://jenkins.scylladb.com/job/sct-github-PRs-scan/job/scylla-dtest/job/PR-4717/32/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/test_hintedhandoff_sync_point_api/

Logs:
[dtest-gw0.log](https://github.com/user-attachments/files/16440472/dtest-gw0.log)
[dtest-gw1.log](https://github.com/user-attachments/files/16440473/dtest-gw1.log)
[dtest-gw2.log](https://github.com/user-attachments/files/16440474/dtest-gw2.log)
[node1.log](https://github.com/user-attachments/files/16440475/node1.log)
[node2.log](https://github.com/user-attachments/files/16440477/node2.log)

The core dump is too large to post it here (60MB+).

**Comment by dawmd** [comment]
I posted a backport of scylladb/scylladb#19134: scylladb/scylladb#19967. It should fix the test. As I described in this issue and in the PR, it's possible that some similar failures may still occur (when someone starts awaiting a sync point when there's no longer a host ID–IP mapping in the locator::token_metadata, the sync point was created before the migration of hinted handoff to host-ID-based, but awaited after it). It's a niche scenario that shouldn't affect any tests. It may affect the user, but it seems unlikely.

Nonetheless, if that problem should be fixed, a new issue should be created for it.
